[teamd] retry creating team_port after interface info changed

yxieca opened the pull request:
**- What I did**

Race condition has been noticed after warm reboot: sometimes when
port_changed notification was received, the link message didn't
have the device name. Without device name, creating team port
would fail.

Registering to the interface information change notification, so
later when device name becomes available, retry creating team port.

Signed-off-by: Ying Xie <ying.xie@microsoft.com>

**- How to verify it**
Continuous warm reboot on my DUT. The retry has been verified with debug messages.

Without the change, continuous warm reboot would fail within 20 iterations. With the fix, the test count has gone up to 78 and still running.

for f in `ls *.log`; do cnt=`grep iteration $f | wc -l`; echo $f $cnt; done
wb-test-20190313-0437.log 13
wb-test-20190319-2212.log 2
wb-test-20190319-2307.log 1
wb-test-20190319-2313.log 3
wb-test-20190319-2348.log 6
wb-test-20190320-0119.log 20
wb-test-20190321-0020.log 2
wb-test-20190321-0207.log 10
wb-test-20190321-1733.log 7
wb-test-20190321-1839.log 5
wb-test-20190321-2131.log 7
wb-test-20190322-0133.log 4
wb-test-20190322-0219.log 5
wb-test-20190322-0429.log 3
wb-test-20190322-1656.log 19
wb-test-20190322-2158.log 2
wb-test-20190323-0006.log 7
wb-test-20190323-0650.log 4
wb-test-20190323-0729.log 3
wb-test-20190323-1855.log 2
wb-test-20190323-1911.log 1
wb-test-20190323-1924.log 10
**wb-test-20190324-1911.log 78**

jipanyang commented:
Could you share the error log collected for the problem?   I don't quite understand how "port_changed notification was received, the link message didn't have the device name" happened.

Also I'd like to know whether it is related to the incorrect teammgrd then teamsyncd startup order.

yxieca commented:
The logs came with libteamd doesn't have all the information. I had to add quite some extra log entries to eventually piece together the puzzle (keep in mind that anything started with '===' was added or enhanced:

1. It all started with lag got created, no member initially:
Mar 25 03:44:07.155421 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]: <ifinfo_list>
Mar 25 03:44:07.155421 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]:  9: PortChannel0004: 74:83:ef:0a:d0:79: 0
Mar 25 03:44:07.155421 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]: </ifinfo_list>
Mar 25 03:44:07.155421 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]: <port_list>
Mar 25 03:44:07.155421 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]: *104: : up 0Mbit HD
Mar 25 03:44:07.155421 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]: </port_list>

2. At some point, the port change was notified, noting that we know the interface index is 104 but no device name yet:
Mar 25 03:44:07.155421 str-7260cx3-acs-1 ERR teamd#teamd_PortChannel0004[47]: === allocating port_obj 0x240b6e0 for team_port 0x2406cf0 name  ifindex 104 ifinfo 0x24206e0

3. That triggers creating lag failure (the first one was added, the second one was enhanced, you should be able to see "Failed to init port priv" in log:
Mar 25 03:44:07.164984 str-7260cx3-acs-1 ERR teamd#teamd_PortChannel0004[47]: === ioctl SIOCADDMULTI failed: dev '' (0x2420750). lp 0x2419740 tp 0x240b6e0
Mar 25 03:44:07.189333 str-7260cx3-acs-1 ERR teamd#teamd_PortChannel0004[47]: === Failed to init port priv. po 0x240b6e0, ppi 0x2419720, ppi->priv 0x2419740, ppt->c_priv 0x2419738

4. Shortly later, you see another event, this time with device name Ethernet72. Without registering for interface change, you don't get this event and the lag won't be created, since there would be no more port change for Ethernet72:
Mar 25 03:44:07.192357 str-7260cx3-acs-1 ERR teamd#teamd_PortChannel0004[47]: === allocating port_obj 0x240b6e0 for team_port 0x2406cf0 name Ethernet72 ifindex 104 ifinfo 0x24206e0

5. Now you have 1 port in lag, the other port came up with proper device name and the lag is then ready (of course the race condition could happen to both port at the same time):

Mar 25 03:44:07.223354 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]: <ifinfo_list>
Mar 25 03:44:07.223354 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]:  104: Ethernet72: 74:83:ef:0a:d0:79: 9
Mar 25 03:44:07.223354 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]:  9: PortChannel0004: 74:83:ef:0a:d0:79: 0
Mar 25 03:44:07.223354 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]: </ifinfo_list>
Mar 25 03:44:07.223354 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]: <port_list>
Mar 25 03:44:07.223354 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]: *107: Ethernet76: up 0Mbit HD
Mar 25 03:44:07.223354 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]:  104: Ethernet72: up 0Mbit HD
Mar 25 03:44:07.223354 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]: </port_list>
Mar 25 03:44:07.223354 str-7260cx3-acs-1 ERR teamd#teamd_PortChannel0004[47]: === allocating port_obj 0x2422e80 for team_port 0x240bbd0 name Ethernet76 ifindex 107 ifinfo 0x240ec70
Mar 25 03:44:07.223354 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]: Ethernet76: Using implicit link watch.
Mar 25 03:44:07.223354 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]: Ethernet76: Got link watch from port config.
Mar 25 03:44:07.223354 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]: Ethernet76: Using lacp_prio "255".
Mar 25 03:44:07.223354 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]: Ethernet76: Using lacp_key "0".
Mar 25 03:44:07.223354 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]: Ethernet76: Using sticky "0".
Mar 25 03:44:07.223354 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]: Added loop callback: lacp_socket, 0x2406e10
Mar 25 03:44:07.223354 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]: Added loop callback: lacp_periodic, 0x2406e10
Mar 25 03:44:07.223354 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]: Ethernet76: Setting periodic timer to "slow".
Mar 25 03:44:07.223354 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]: Added loop callback: lacp_timeout, 0x2406e10
Mar 25 03:44:07.223882 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]: <ifinfo_list>
Mar 25 03:44:07.223882 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]:  107: Ethernet76: 74:83:ef:0a:d0:79: 9
Mar 25 03:44:07.223882 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]:  104: Ethernet72: 74:83:ef:0a:d0:79: 9
Mar 25 03:44:07.223882 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]:  9: PortChannel0004: 74:83:ef:0a:d0:79: 0
Mar 25 03:44:07.223882 str-7260cx3-acs-1 DEBUG teamd#teamd_PortChannel0004[47]: </ifinfo_list>

yxieca commented:
retest this please

jipanyang commented:
Currently TeamMgr::doLagMemberTask() makes sure member port state is ok before adding the port to LAG.   Port state ok was set by portsyncd upon receiving  RTM_NEWLINK message which contains the correct device name.

Why when the processing reaches teamd later, device/link name becomes unavailable?  bug in teamd?

yxieca commented:
I had the same question, but with the information from you question, I think I have an answer now:

1. It is a known issue that initial RTM_NEWLINK could come up without device name.
2. Teamdmgr dealt with it by adding a wait until device name is available.
3. Similar thing were now done in teamd.

Why we need to do it again? Because after warm reboot, the LAGs are restored by teamd. This operation bypassed the teammgrd protection.

So I guess we really need to fix it in libnl, I looked at libnl while investigating this issue. I didn't find a good way of patching libnl. Maybe someone more familiar with libnl could take another look?

Cheers,
Ying

yxieca commented:
Overnight test reached 172 iterations, all 4 peers see lags up for ~21 hours.

yinxi@acs-trusty8:~$ ssh vm-1 "show interface po1" 
Password: 
Port-Channel1 is up, line protocol is up (connected)
  Hardware is Port-Channel, address is 5254.003f.38de
  Internet address is 10.0.0.33/31
  Broadcast address is 255.255.255.255
  Address determined by manual configuration
  IPv6 link-local address is fe80::5054:ff:fe3f:38de/64
  IPv6 global unicast address(es):
    fc00::22, subnet is fc00::20/126
  IP MTU 1500 bytes
  Full-duplex, Unconfigured 
  Active members in this channel: 2
  ... Ethernet1 , Full-duplex, Unconfigured
  ... Ethernet2 , Full-duplex, Unconfigured
  Fallback mode is: off
  Up 20 hours, 51 minutes, 7 seconds
  127 link status changes since last clear
  Last clearing of "show interface" counters never
  5 minutes input rate 8.75 kbps (- with framing overhead), 3 packets/sec
  5 minutes output rate 20.7 kbps (- with framing overhead), 4 packets/sec
     30029550 packets input, 4372447368 bytes
     Received 4477 broadcasts, 589101 multicast
     0 input errors, 0 input discards
     11082681 packets output, 3004702653 bytes
     Sent 514 broadcasts, 525686 multicast
     0 output errors, 0 output discards
yinxi@acs-trusty8:~$ ssh vm-2 "show interface po1" 
Password: 
Port-Channel1 is up, line protocol is up (connected)
  Hardware is Port-Channel, address is 5254.0046.c8e3
  Internet address is 10.0.0.35/31
  Broadcast address is 255.255.255.255
  Address determined by manual configuration
  IPv6 link-local address is fe80::5054:ff:fe46:c8e3/64
  IPv6 global unicast address(es):
    fc00::26, subnet is fc00::24/126
  IP MTU 1500 bytes
  Full-duplex, Unconfigured 
  Active members in this channel: 2
  ... Ethernet1 , Full-duplex, Unconfigured
  ... Ethernet2 , Full-duplex, Unconfigured
  Fallback mode is: off
  Up 20 hours, 51 minutes, 15 seconds
  668 link status changes since last clear
  Last clearing of "show interface" counters never
  5 minutes input rate 9.49 kbps (- with framing overhead), 3 packets/sec
  5 minutes output rate 20.2 kbps (- with framing overhead), 4 packets/sec
     45463370 packets input, 7356854288 bytes
     Received 6384 broadcasts, 920840 multicast
     0 input errors, 0 input discards
     17286065 packets output, 5228048304 bytes
     Sent 2327 broadcasts, 839577 multicast
     0 output errors, 0 output discards
yinxi@acs-trusty8:~$ ssh vm-3 "show interface po1" 
Password: 
Port-Channel1 is up, line protocol is up (connected)
  Hardware is Port-Channel, address is 5254.0053.0dbf
  Internet address is 10.0.0.37/31
  Broadcast address is 255.255.255.255
  Address determined by manual configuration
  IPv6 link-local address is fe80::5054:ff:fe53:dbf/64
  IPv6 global unicast address(es):
    fc00::2a, subnet is fc00::28/126
  IP MTU 1500 bytes
  Full-duplex, Unconfigured 
  Active members in this channel: 2
  ... Ethernet1 , Full-duplex, Unconfigured
  ... Ethernet2 , Full-duplex, Unconfigured
  Fallback mode is: off
  Up 20 hours, 51 minutes, 37 seconds
  756 link status changes since last clear
  Last clearing of "show interface" counters never
  5 minutes input rate 23.8 kbps (- with framing overhead), 10 packets/sec
  5 minutes output rate 19.6 kbps (- with framing overhead), 3 packets/sec
     30547304 packets input, 5701239147 bytes
     Received 6296 broadcasts, 908131 multicast
     0 input errors, 0 input discards
     14157728 packets output, 4789280043 bytes
     Sent 3299 broadcasts, 822843 multicast
     0 output errors, 0 output discards
yinxi@acs-trusty8:~$ ssh vm-4 "show interface po1" 
Password: 
Port-Channel1 is up, line protocol is up (connected)
  Hardware is Port-Channel, address is 5254.00bb.81fd
  Internet address is 10.0.0.39/31
  Broadcast address is 255.255.255.255
  Address determined by manual configuration
  IPv6 link-local address is fe80::5054:ff:febb:81fd/64
  IPv6 global unicast address(es):
    fc00::2e, subnet is fc00::2c/126
  IP MTU 1500 bytes
  Full-duplex, Unconfigured 
  Active members in this channel: 2
  ... Ethernet1 , Full-duplex, Unconfigured
  ... Ethernet2 , Full-duplex, Unconfigured
  Fallback mode is: off
  Up 20 hours, 51 minutes, 39 seconds
  768 link status changes since last clear
  Last clearing of "show interface" counters never
  5 minutes input rate 21.1 kbps (- with framing overhead), 3 packets/sec
  5 minutes output rate 18.8 kbps (- with framing overhead), 3 packets/sec
     27506835 packets input, 5774821693 bytes
     Received 6208 broadcasts, 741207 multicast
     0 input errors, 0 input discards
     10433764 packets output, 4298133173 bytes
     Sent 3309 broadcasts, 640459 multicast
     0 output errors, 0 output discards

jipanyang commented:
"It is a known issue that initial RTM_NEWLINK could come up without device name.
Teamdmgr dealt with it by adding a wait until device name is available.
"   Could you help point me the the issue/code change? TeamMgr::addLagMember() doesn't seem to have the exact processing.

For system warm reboot, the kernel is clean, though lacp state will be restored from the saved lacp pdu file, LAG and lag member are managed by teammmgrd.    How this operation bypassed the teammgrd protection?

Your change looks fixing the problem. I just want to get a clear understanding of the problem scenario

yxieca commented:
Things are getting more interesting. The libnl has an 'optimization' that it won't report interface information change if it has been changed once. Looks like we might be hitting that too.

At 182th iteration, test hit an issue where teamd repeated retry creating port_obj with empty name. Now we need even longer test cycle to proof safety.

yxieca commented:
There is a change in teamd to behave differently under warm reboot. Because we need to keep the lag member ID consistent across warm reboot so the neighbor wouldn't be confused and tear down the lag.

yxieca commented:
Jipan,

I saw some change under warm reboot, teamd would do things differently. That might be mainly to make sure that we use same lag member ID to prevent peer tearing down the LAGs. 

What you claimed might be right. I didn't fully track down who created the new LAGs. One thing I know is that the LAGs are created very early on after warm reboot, they are usually already created by the time I can ssh in.

Cheers,
Ying
